File: windows/GestorFinances.Tests/Domain/Rules/DebtConsumption.cs
namespace GestorFinances.Domain.Rules;

/// <summary>Which debt a settlement is allowed to consume.</summary>
public enum SettlementScope
{
    All,
    Recurring
}

/// <summary>
/// One row of a person's balance breakdown, in v_person_balance signs: a positive
/// <paramref name="EffectCents"/> means the person owes the user more. <paramref name="IsRecurring"/>
/// marks debt originating from a recurring template, which is what a Recurring-scoped settlement is
/// allowed to consume. <paramref name="Scope"/> is set on settlements only.
/// </summary>
public sealed record DebtItem(
    string SourceId,
    string Date,
    long EffectCents,
    bool IsSettlement,
    bool IsRecurring,
    SettlementScope? Scope);

/// <summary>A debt item that settlements have not fully consumed.</summary>
public sealed record DebtResidual(string SourceId, long OriginalCents, long RemainingCents)
{
    /// <summary>True when settlements consumed part but not all of this item.</summary>
    public bool IsPartial => RemainingCents != OriginalCents;
}

/// <summary>
/// What still explains a person's balance, and the credit no debt has absorbed yet.
/// <paramref name="TotalCents"/> is the sum of every item's effect, exactly what v_person_balance
/// derives; residuals plus both credit buckets always reconcile to it.
/// </summary>
public sealed record DebtProjection(
    IReadOnlyList<DebtResidual> Residuals,
    long CreditAllCents,
    long CreditRecurringCents,
    long TotalCents);

/// <summary>
/// Explains a person's balance by consuming debt chronologically, oldest first.
///
/// A settlement applies only to eligible debt that already exists on its date, so items are folded
/// in date order with debt ranked before settlements on the same day. Whatever a settlement cannot
/// spend stays as directional credit in its own scope bucket and is absorbed by later debt, which
/// keeps a recurring-scoped overpayment from silently paying off unrelated expenses.
///
/// Locked by shared/golden/debt_consumption.json.
/// </summary>
public static class DebtConsumption
{
    public static DebtProjection Project(IReadOnlyList<DebtItem> items)
    {
        var ordered = items
            .OrderBy(item => item.Date, StringComparer.Ordinal)
            .ThenBy(item => item.IsSettlement ? 1 : 0)
            .ThenBy(item => item.SourceId, StringComparer.Ordinal)
            .ToList();

        var open = new List<OpenItem>();
        var creditAll = 0L;
        var creditRecurring = 0L;

        foreach (var item in ordered)
        {
            if (item.EffectCents == 0)
            {
                continue;
            }

            if (item.IsSettlement)
            {
                var scope = item.Scope ?? throw new InvalidOperationException(
                    $"Settlement {item.SourceId} has no scope.");
                var power = Math.Abs(item.EffectCents);
                var direction = Math.Sign(item.EffectCents);

                foreach (var candidate in open)
                {
                    if (power == 0)
                    {
                        break;
                    }

                    if (Math.Sign(candidate.Remaining) != -direction)
                    {
                        continue;
                    }

                    if (scope == SettlementScope.Recurring && !candidate.IsRecurring)
                    {
                        continue;
                    }

                    var take = Math.Min(power, Math.Abs(candidate.Remaining));
                    candidate.Remaining += direction * take;
                    power -= take;
                }

                if (power > 0)
                {
                    var leftover = direction * power;
                    if (scope == SettlementScope.All)
                    {
                        creditAll += leftover;
                    }
                    else
                    {
                        creditRecurring += leftover;
                    }
                }
            }
            else
            {
                var remaining = item.EffectCents;

                // Spend the narrower recurring credit first so it cannot be stranded by an
                // 'all' credit that any later item could have absorbed instead.
                if (item.IsRecurring)
                {
                    (remaining, creditRecurring) = Absorb(remaining, creditRecurring);
                }

                (remaining, creditAll) = Absorb(remaining, creditAll);

                if (remaining != 0)
                {
                    open.Add(new OpenItem(item.SourceId, item.EffectCents, remaining, item.IsRecurring));
                }
            }
        }

        var residuals = open
            .Where(row => row.Remaining != 0)
            .Select(row => new DebtResidual(row.SourceId, row.OriginalCents, row.Remaining))
            .ToList();

        return new DebtProjection(
            residuals,
            creditAll,
            creditRecurring,
            items.Sum(item => item.EffectCents));
    }

    /// <summary>Applies opposite-signed credit to an amount, returning both after the transfer.</summary>
    private static (long Amount, long Credit) Absorb(long amount, long credit)
    {
        if (amount == 0 || credit == 0 || Math.Sign(credit) != -Math.Sign(amount))
        {
            return (amount, credit);
        }

        var take = Math.Min(Math.Abs(amount), Math.Abs(credit));
        return (amount + Math.Sign(credit) * take, credit - Math.Sign(credit) * take);
    }

    private sealed class OpenItem(string sourceId, long originalCents, long remaining, bool isRecurring)
    {
        public string SourceId { get; } = sourceId;

        public long OriginalCents { get; } = originalCents;

        public long Remaining { get; set; } = remaining;

        public bool IsRecurring { get; } = isRecurring;
    }
}

File: android/app/src/main/java/com/gestorfinances/app/domain/rules/DebtConsumption.kt
package com.gestorfinances.app.domain.rules

/** Which debt a settlement is allowed to consume. */
enum class SettlementScope(val dbValue: String) {
    ALL("all"),
    RECURRING("recurring"),
    ;

    companion object {
        fun fromDb(value: String): SettlementScope =
            entries.firstOrNull { it.dbValue == value } ?: error("Unknown settlement scope: $value")
    }
}

/**
 * One row of a person's balance breakdown, in `v_person_balance` signs: a positive [effectCents]
 * means the person owes the user more.
 *
 * [isRecurring] marks debt that originated from a recurring template, which is what a
 * `RECURRING`-scoped settlement is allowed to consume. [scope] is set on settlements only.
 */
data class DebtItem(
    val sourceId: String,
    val date: String,
    val effectCents: Long,
    val isSettlement: Boolean,
    val isRecurring: Boolean,
    val scope: SettlementScope?,
)

/** A debt item that settlements have not fully consumed. */
data class DebtResidual(
    val sourceId: String,
    val originalCents: Long,
    val remainingCents: Long,
) {
    /** True when settlements consumed part but not all of this item. */
    val isPartial: Boolean get() = remainingCents != originalCents
}

/**
 * What still explains a person's balance, and the credit that no debt has absorbed yet.
 *
 * [totalCents] is the sum of every item's effect, which is exactly what `v_person_balance` derives.
 * The residuals plus both credit buckets always reconcile to it.
 */
data class DebtProjection(
    val residuals: List<DebtResidual>,
    val creditAllCents: Long,
    val creditRecurringCents: Long,
    val totalCents: Long,
)

/**
 * Explains a person's balance by consuming debt chronologically, oldest first.
 *
 * A settlement applies only to eligible debt that already exists on its date, so items are folded
 * in date order with debt ranked before settlements on the same day. Whatever a settlement cannot
 * spend stays as directional credit in its own scope bucket and is absorbed by later debt, which
 * keeps a recurring-scoped overpayment from silently paying off unrelated expenses.
 *
 * Locked by `shared/golden/debt_consumption.json`.
 */
object DebtConsumption {
    fun project(items: List<DebtItem>): DebtProjection {
        val ordered = items.sortedWith(
            compareBy({ it.date }, { if (it.isSettlement) 1 else 0 }, { it.sourceId }),
        )
        val open = mutableListOf<OpenItem>()
        var creditAll = 0L
        var creditRecurring = 0L

        for (item in ordered) {
            if (item.effectCents == 0L) continue
            if (item.isSettlement) {
                val scope = requireNotNull(item.scope) { "Settlement ${item.sourceId} has no scope." }
                var power = kotlin.math.abs(item.effectCents)
                val direction = item.effectCents.sign()
                for (candidate in open) {
                    if (power == 0L) break
                    if (candidate.remaining.sign() != -direction) continue
                    if (scope == SettlementScope.RECURRING && !candidate.isRecurring) continue
                    val take = minOf(power, kotlin.math.abs(candidate.remaining))
                    candidate.remaining += direction * take
                    power -= take
                }
                if (power > 0L) {
                    val leftover = direction * power
                    when (scope) {
                        SettlementScope.ALL -> creditAll += leftover
                        SettlementScope.RECURRING -> creditRecurring += leftover
                    }
                }
            } else {
                var remaining = item.effectCents
                // Spend the narrower recurring credit first so it cannot be stranded by an
                // 'all' credit that any later item could have absorbed instead.
                if (item.isRecurring) {
                    val (left, credit) = absorb(remaining, creditRecurring)
                    remaining = left
                    creditRecurring = credit
                }
                val (left, credit) = absorb(remaining, creditAll)
                remaining = left
                creditAll = credit
                if (remaining != 0L) {
                    open += OpenItem(item.sourceId, item.effectCents, remaining, item.isRecurring)
                }
            }
        }

        return DebtProjection(
            residuals = open
                .filter { it.remaining != 0L }
                .map { DebtResidual(it.sourceId, it.originalCents, it.remaining) },
            creditAllCents = creditAll,
            creditRecurringCents = creditRecurring,
            totalCents = items.sumOf { it.effectCents },
        )
    }

    /** Applies opposite-signed [credit] to [amount], returning both after the transfer. */
    private fun absorb(amount: Long, credit: Long): Pair<Long, Long> {
        if (amount == 0L || credit == 0L || credit.sign() != -amount.sign()) return amount to credit
        val take = minOf(kotlin.math.abs(amount), kotlin.math.abs(credit))
        return (amount + credit.sign() * take) to (credit - credit.sign() * take)
    }

    private fun Long.sign(): Long = when {
        this > 0L -> 1L
        this < 0L -> -1L
        else -> 0L
    }

    private class OpenItem(
        val sourceId: String,
        val originalCents: Long,
        var remaining: Long,
        val isRecurring: Boolean,
    )
}
